runtime/drivers/pinot/driver.go

```diff
@@ -0,0 +1,282 @@
+package pinot
+
+import (
+	"context"
+	"database/sql"
+	sqlDriver "database/sql/driver"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"math/big"
+	"net/url"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/startreedata/pinot-client-go/pinot"
+)
+
+type pinotDriver struct{}
+
+func (d *pinotDriver) Open(dsn string) (sqlDriver.Conn, error) {
+	// validate dsn - it should be a valid URL, may contain basic auth credentials
+	u, err := url.Parse(dsn)
+	if err != nil {
+		return nil, fmt.Errorf("invalid DSN: %w", err)
+	}
+
+	var authHeader map[string]string
+	if u.User != nil {
+		uname := u.User.Username()
+		pwd, passwordSet := u.User.Password()
+		if uname == "" || !passwordSet {
+			return nil, fmt.Errorf("DSN should contain valid basic auth credentials")
+		}
+		// clear user info from URL so that u.String() doesn't include it
+		u.User = nil
+		authString := fmt.Sprintf("%s:%s", uname, pwd)
+		authHeader = map[string]string{
+			"Authorization": fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(authString))),
+		}
+	}
+
+	pinotConn, err := pinot.NewWithConfig(&pinot.ClientConfig{
+		ExtraHTTPHeader: authHeader,
+		ControllerConfig: &pinot.ControllerConfig{
+			ControllerAddress: u.String(),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	// TODO check if it needs to be configurable
+	pinotConn.UseMultistageEngine(true)
```

Comment:
I think we would not want to make this configurable, as joins and nested queries are only supported on multi stage engine which is off by default so really confusing. 

However, single stage engine is more effecient, has more intuitive response types and simple semantics to work with multi valued columns. Working with timestamps column on both engine vary for example with MSE I can do `ts > '2014-01-02 00:00:00.000Z'` but on SSE 'Z' is not supported, similarly I can cast timestamp column to long (millis) in SSE but it throws error with MSE, you have to cast to double. Anyways since we need joins and nested queries, we can just keep this true. I can modify the comment.

Reply:
Makes sense